File: 2018/ruby/day02.rb
input = File.readlines("/Users/hhh/JungleGym/advent_of_code/2018/inputs/day2.in").map { |x| x. strip }
twos = 0
threes = 0

input.each do |x|
  counts = Hash.new 0

  x.chars.each do |char|
    counts[char] += 1
  end

  unique_counts = []
  counts.each do |char, count|
    unique_counts << count
  end
  unique_counts.uniq!

  twos += 1 if unique_counts.include? 2
  threes += 1 if unique_counts.include? 3
end

p "Part 1: #{twos * threes}"

match = []
input.each do |x|
  input.each do |y|
    pairs = x.chars.zip(y.chars)
    diffs = pairs.select { |a, b| a != b }
    match = [x, y] if diffs.count == 1
  end
end

letters = match.first.chars.zip(match.last.chars).select  {|a, b| a == b }.map { |a, b| a }.join
p "Part 2: #{letters}"

File: 2018/go/day02/day02.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

func Checksum(ids []string) int {
	twos, threes := 0, 0
	for _, id := range ids {
		// store character counts
		count := make(map[rune]int)
		for _, char := range id {
			count[char]++
		}

		occurences := make(map[int]bool)
		for _, val := range count {
			occurences[val] = true
		}

		if occurences[2] {
			twos++
		}
		if occurences[3] {
			threes++
		}

	}
	return twos * threes
}

func HamOne(ids []string) string {
	for i, id := range ids {
		for j, id2 := range ids {
			diff := 0
			letters := ""

			if j <= i {
				continue
			}

			if len(id) == len(id2) {
				for idx, char := range id {
					if string(char) != string(id2[idx]) {
						diff++
					} else {
						letters += string(char)
					}
				}

			}

			if diff == 1 {
				return letters
			}
		}

	}

	return ""
}


func main() {
	test_ids := []string{
		"abcdef",
		"bababc",
		"abbcde",
		"abcccd",
		"aabcdd",
		"abcdee",
		"ababab",
	}
	fmt.Printf("Test: Expect 12, Got: %+v\n", Checksum(test_ids))

	input, err := ioutil.ReadFile("/Users/hhh/JungleGym/advent_of_code/2018/inputs/day2.in")
	if err != nil {
		panic(err)
	}
	ids := formatInput(string(input))

	fmt.Printf("Day 2 Part 1: %+v\n", Checksum(ids))
	fmt.Printf("Day 2 Part 2: %+v\n", HamOne(ids))
}

func formatInput(input string) []string{
	return strings.Split(input, "\n")
}
